### PR TITLE
Organize solar related things.

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [x] Tests for the changes have been added (for bug fixes / features)
  - [ ] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [ ] HISTORY.rst has been updated (with summary of main changes)
  - [ ] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added
- [ ] `bumpversion patch` has been called on this branch
- [x] The relevant author information has been added to `.zenodo.json`

### What kind of change does this PR introduce?
So I wanted to help out on `utci` and make all those solar angles and radiation calculations more organized and accessible and the change grew so large that I decided to make a PR so we can discuss this separately from the UTCI stuff.

* I reorganized all solar angle and radiation functions into `_helpers.py`. This file is meant of helper functions that manipulate real data and perform some physical operations. (In opposition to `generic` which is more "meta").
* I used citable equations. This meant modifying some calculations, at least the default version. This was easy going for the extraterrestrial radiation of the potential evapotranspiration, but not for BEDD and Huglin. I thus had to go back an reimplement the "old" ways and add a "method" arg to choose between equations.
* As `latitude` is needed for all algorithms of PET, I added it as an explicit input. We could use `cf_xarray`, but I'm not sure how generic this is.

### Does this PR introduce a breaking change?
`lat` is now needed for `potential_evapotranspiration` and `water_budget`.

### Other information:
@ludwiglierhammer When this is ready, you will be able to use `cosine_of_solar_zenith_angle` in the `Tmrt` function.

The issues with precision are because of two functions:
1. Solar declination angle
    The new method is closest to reality, it uses a few terms of the fourier series of the observations. The old method assumed that the orbit is a circle, which is good enough, but you know.
2. Eccentricity correction factor
    Similar: the new method uses the first terms of the fourier series of the obs. The old only used the 2 first terms.

@Zeitsperre  How strict do we need to be with the tests of BEDD and Huglin? The problem is that the very small difference in day length that the different solar declination equation yields adds up over the aggregation period and we get relative errors of around 0.01, which is larger than the current tolerance.

@raquel-ucl Flagging you so you are aware of the changes proposed here. They will slightly affect results of the MB05 methods you guys added in xclim.

Over all (@huard @tlogan2000 ), am I being too verbose here? I thought this would be simple and now I added multiple lines of code for functions that could have been implicit (but duplicated).

### CODE REVIEW

**Comment by Zeitsperre** [comment]
@aulemahal I don't think it's incredibly important for those indices. I can't remember the source I used for the day-length factor calculations (was it Jones?), but so long as the references are present and the docstring clearly indicates the formula and source, I think it's fine.

Alternatively, would it be possible to move the old method into a new function? If the effort required is too much for the benefit, I can try adding it in myself later. The only real benefit would be to replicate/validate older studies that used it, I suppose.

**Comment by aulemahal** [comment]
Good! The day-length you took from Jones and I'm guessing they took it from elsewhere, but this the kind of equation that doesn't really need (IMO) an explicit citation; it really just is orbital mechanics. The set of assumptions, though, could be attributed to Jones here. However, my main concern for BEDD and Huglin is that the "jones" method is only very slightly different than the "simple" one (which is, to my knowledge, more commonly used). The more exact "spencer" method does give a larger difference.  ( I'm thinking changing the "jones" name for something more explicit, but haven't found a good one yet) 

The current code uses the same method you coded, it was simply moved to its own function with the classic `method` arg for switching algos. However, even if multiple versions are coded, I did _not_ propagate the switch argument up to the indice. Currently, BEDD and Huglin will use "jones" to compute solar declination (for day lengths) while PET will use "spencer".

While removing "jones" if favor of "simple" would be my favorite option, it's ok if we keep it that way for more reproducibility. It's a duplication of pathways, but  (to me) that's still better than duplication of code.

File: xclim/indices/_helpers.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    day_angle: xr.DataArray
+        Assuming the earth makes a full circle in a year, this is the angle covered from
+        the beginning of the year up to that timestep. Also called the "julian day fraction".
+        See :py:func:`~xclim.core.calendar.datetime_to_decimal_year`.
+    method: {'spencer', 'simple', 'jones'}
+        Which approximation to use. The default ("spencer") uses the first 7 terms of the
+        Fourier series representing the observed declination.
+        The two other methods assume the orbit is a circle with a fixed obliquity.
+        "simple" assumes the solstice/equinox happen at fixed angles on the orbit (the date changes for leap years).
+        "jones" assumes the solcstice/equinox happen at fixed dates (the angle changes for leap years).
```
Comment (huard) [inline xclim/indices/_helpers.py]:
```suggestion
        "jones" assumes the solstice/equinox happen at fixed dates (the angle changes for leap years).
```

File: xclim/indices/_helpers.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    Solar declination angle, [rad]
+
+    Reference
+    ---------
+    .. [Spencer1971] Spencer JW (1971) Fourier series representation of the position of the sun. Search 2(5):172
+    """
+    # julian day fraction
+    da = convert_units_to(day_angle, "rad")
+    if method == "simple":
+        # Another approximation is
```
Comment (huard) [inline xclim/indices/_helpers.py]:
```suggestion

```

File: xclim/indices/_helpers.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    """Extraterrestrial solar radiation
+
+    This is the daily energy received on a surface parallel to the ground at the mean
+    distance of the earth to the sun. It neglects the effect of the atmosphere. Computation
+    is based on [Kalogirou14]_ and the default solar constant is taken from [Matthes17]_.
+
+    Parameters
+    ----------
+    times: xr.DataArray
+            Daily datetime data. This function makes no sense with data of other frequency.
```
Comment (huard) [inline xclim/indices/_helpers.py]:
```suggestion
        Daily datetime data. This function makes no sense with data of other frequency.
```
Worth checking the frequency in this case, and raise an Error if not daily ?

File: xclim/indices/_helpers.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+    Parameters
+    ----------
+    dates: xr.DataArray
+    lat: xarray.DataArray
+      Latitude coordinate.
+    start_date: xarray.DataArray or DayOfYearStr, optional
+      Start date to consider for calculating mean day lengths. Default: None.
+    end_date: xarray.DataArray or DayOfYearStr, optional
+      End date to consider for calculating mean day lengths. Default: None.
```
Comment (huard) [inline xclim/indices/_helpers.py]: Is that necessary? If that's a common use case, an option would be to add an example in the docstring.

**Review by huard — APPROVED** [review]
Are you counting on indicator tests to validate the solar functions? I think it would still be useful to test the individual functions, at least for edge cases.

File: xclim/indices/_helpers.py (hunk 4 above)
Comment (Zeitsperre) [reply xclim/indices/_helpers.py]: Yes, the start and end dates are needed for agroclimate indices. An example of an application: doi:10.1016/j.agrformet.2003.06.001

Comment (aulemahal) [reply xclim/indices/_helpers.py]: @Zeitsperre I thought about it and maybe this is not a use case common enough to be implemented directly in the `day_lengths` function? My next commit moves the aggregation to the BEDD and Huglin indices.

File: xclim/indices/_helpers.py (hunk 3 above)
Comment (aulemahal) [reply xclim/indices/_helpers.py]: We never do that for indices, so I don't think we should do it for helper function. Yes user may receive wrong data, but if they used a helper function directly instead of an indicator, I assume they made the choice to "bypass" these checks.

**Comment by aulemahal** [comment]
@huard  I only added tests for the functions where I could find an online calculator. I was lazy sorry, it's friday :). The indices/indicator tests will cover for the rest.
@ludwiglierhammer I am currently running the full test suite, I'll ping you monday the latest if it passes and you'll be able to merge this and continue development of UTCI.

**Comment by aulemahal** [comment]
@ludwiglierhammer  Sorry for the delay. The last commit fixed remaning errors in the tests, you can now merge this branch into yours and continue development on your PR!
